Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,782 @@
+    [Parameter(ParameterSetName = 'EnableAMSIAll', Mandatory = $true)]
     [switch]$EnableAMSI,
-    [Parameter(ParameterSetName = 'DisableAMSI', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'DisableAMSI', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'DisableAMSIAll', Mandatory = $true)]
     [switch]$DisableAMSI,
+    [Parameter(ParameterSetName = 'RestartIIS', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $true)]
+    [switch]$RestartIIS,
+    [Alias("ExchangeServerFQDN")]
+    [Parameter(ParameterSetName = 'TestAMSI', Mandatory = $false, ValueFromPipeline = $true)]
+    [Parameter(ParameterSetName = 'EnableAMSI', Mandatory = $false, ValueFromPipeline = $true)]
+    [Parameter(ParameterSetName = 'DisableAMSI', Mandatory = $false, ValueFromPipeline = $true)]
+    [Parameter(ParameterSetName = 'CheckAMSIConfig', Mandatory = $false, ValueFromPipeline = $true)]
+    [Parameter(ParameterSetName = 'RestartIIS', Mandatory = $false, ValueFromPipeline = $true)]
+    [string[]]$ServerList = $null,
+    [Parameter(ParameterSetName = 'TestAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'CheckAMSIConfigAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'EnableAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'DisableAMSIAll', Mandatory = $true)]
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $true)]
+    [switch]$AllServers,
+    [Parameter(ParameterSetName = 'TestAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'CheckAMSIConfigAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'EnableAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'DisableAMSIAll', Mandatory = $false)]
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $false)]
+    [string[]]$Sites = $null,
     [Parameter(ParameterSetName = 'RestartIIS', Mandatory = $false)]
-    [switch]$RestartIIS
+    [Parameter(ParameterSetName = 'RestartIISAll', Mandatory = $false)]
+    [switch]$Force
 )
 
-function Confirm-Administrator {
-    $currentPrincipal = New-Object Security.Principal.WindowsPrincipal( [Security.Principal.WindowsIdentity]::GetCurrent() )
-    if ($currentPrincipal.IsInRole( [Security.Principal.WindowsBuiltInRole]::Administrator )) {
-        return $true
-    } else {
-        return $false
-    }
-}
+begin {
 
-function SetCertificateValidationBehavior {
-    [CmdletBinding()]
-    param (
-        [Parameter(ParameterSetName = "Ignore", Mandatory = $true)]
-        [switch]
-        $Ignore,
-
-        [Parameter(ParameterSetName = "Default", Mandatory = $true)]
-        [switch]
-        $Default
-    )
-
-    if ($Ignore) {
-        Add-Type @"
-        using System;
-        using System.Net;
-        using System.Net.Security;
-        using System.Security.Cryptography.X509Certificates;
-        public class ServerCertificateValidationBehavior
+    Add-Type @"
```

Comment:
We have the `Enable-TrustAnyCertificateCallback.ps1` shared function (can be found here: `Shared/CertificateFunctions/Enable-TrustAnyCertificateCallback.ps1`) which could be used here.